File: src/TidyTree/Node.ts
// Import Types
import type { TidyInfo } from "./TidyInfo";

export class Node {
  id: number;
  width: number;
  height: number;
  x: number;
  y: number;
  index: number;
  relative_x: number;
  relative_y: number;
  bounding_box_w: number;
  parent?: Node;
  children: Array<Node>;
  tidy?: TidyInfo;

  constructor(id: number, w: number = 0, h: number = 0, x: number = 0, y: number = 0) {
    this.id = id;
    this.width = w;
    this.height = h;
    this.x = x;
    this.y = y;
    this.index = 0;
    this.relative_x = 0;
    this.relative_y = 0;
    this.bounding_box_w = 0;
    this.parent = undefined;
    this.children = [];
    this.tidy = undefined;
  }

  clone() {
    let node = new Node(this.id);
    node.width = this.width;
    node.height = this.height;
    node.x = this.x;
    node.y = this.y;
    node.index = this.index;
    node.relative_x = this.relative_x;
    node.relative_y = this.relative_y;
    node.bounding_box_w = this.bounding_box_w;
    node.parent = this.parent;
    node.children = this.children;
    node.tidy = this.tidy;
    return node;
  }

  depth() {
    let depth = 0;
    let node: Node = this;

    while (node.parent) {
      node = node.parent;
      depth += 1;
    }

    return depth;
  }

  bottom() {
    return this.height + this.y;
  }

  append_child(node: Node) {
    node.parent = this;
    this.children.push(node);
  }

  intersects(other: Node) {
    return (
      this.x - this.width / 2 < other.x + other.width / 2 &&
      this.x + this.width / 2 > other.x - other.width / 2 &&
      this.y < other.y + other.height &&
      this.y + this.height > other.y
    );
  }

  remove_child(id: number) {
    let index = this.children.findIndex(x => x.id === id);

    if (index !== -1) {
      this.children.splice(index, 1);
    }
  }
}

File: rust/creates/tidy-tree/src/node.rs
// use rust std
use std::cell::RefCell;
use std::rc::{Rc, Weak};

pub type NodeTupleData = (i64, f32, f32, f32, f32);

pub struct TidyInfo {
    pub thread_left: Weak<RefCell<Node>>,
    pub thread_right: Weak<RefCell<Node>>,
    pub extreme_left: Weak<RefCell<Node>>,
    pub extreme_right: Weak<RefCell<Node>>,
    pub shift_acceleration: f32,
    pub shift_change: f32,
    pub modifier_to_subtree: f32,
    pub modifier_thread_left: f32,
    pub modifier_thread_right: f32,
    pub modifier_extreme_left: f32,
    pub modifier_extreme_right: f32,
    // to accelerate the speed of second walk
    pub prev_modified_sum: f32,
}

impl TidyInfo {
    pub fn new() -> TidyInfo {
        TidyInfo {
            thread_left: Weak::new(),
            thread_right: Weak::new(),
            extreme_left: Weak::new(),
            extreme_right: Weak::new(),
            shift_acceleration: 0.0,
            shift_change: 0.0,
            modifier_to_subtree: 0.0,
            modifier_thread_left: 0.0,
            modifier_thread_right: 0.0,
            modifier_extreme_left: 0.0,
            modifier_extreme_right: 0.0,
            // to accelerate the speed of second walk
            prev_modified_sum: 0.0,
        }
    }
}

pub enum NodeType {
    NORMAL,
    LITE,
    BATCH,
    EXTEND,
}

pub struct Node {
    pub id: i64,
    pub width: f32,
    pub height: f32,
    pub x: f32,
    pub y: f32,
    pub index: usize,
    pub relative_x: f32,
    pub relative_y: f32,
    pub bounding_box_w: f32,
    pub parent: Weak<RefCell<Node>>,
    pub children: Vec<Rc<RefCell<Node>>>,
    pub mode: NodeType,
    pub tidy: Option<TidyInfo>,
}

impl Node {
    pub fn new(id: i64, w: f32, h: f32, mode: NodeType) -> Node {
        Node {
            id,
            children: Vec::new(),
            parent: Weak::new(),
            width: w,
            height: h,
            x: 0.0,
            y: 0.0,
            index: 0,
            relative_x: 0.0,
            relative_y: 0.0,
            bounding_box_w: 0.0,
            mode,
            tidy: None,
        }
    }

    pub fn bottom(&self) -> f32 {
        self.height + self.y
    }

    pub fn to_array(&self) -> (i64, f32, f32, f32, f32) {
        (self.id, self.x, self.y, self.width, self.height)
    }
}
